## Week 2: Exploratory Data Analysis and Baseline Models

- Dataset: Telco Customer Churn, 7043 rows and 21 columns
- EDA Finding: Customers with low tenure and high MonthlyCharges have higher churn. Fiber optic internet users churn more.
- Preprocessing: Converted TotalCharges to numeric, filled 11 missing values with 0, used OneHotEncoder and StandardScaler in a Pipeline.
- Model Performance (80/20 split):
    - Logistic Regression: Accuracy 78.54%, AUC 0.84, Recall 0.52
    - Random Forest: Accuracy 77.92%, AUC 0.82, Recall 0.48
- Best Model: Logistic Regression for better recall and AUC.
- Biggest Lesson: Accuracy is misleading on imbalanced data, recall and AUC are more important for churn prediction.

In [6]:
import pandas as pd
import kagglehub
from kagglehub import KaggleDatasetAdapter
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

file_path = "WA_Fn-UseC_-Telco-Customer-Churn.csv"
df = kagglehub.load_dataset(
  KaggleDatasetAdapter.PANDAS,
  "blastchar/telco-customer-churn",
  file_path,
)
print("Data Load Ho Gya! Shape:", df.shape)

# Fix 1: dropna ki jagah fillna
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
df['TotalCharges'] = df['TotalCharges'].fillna(0)

# Fix 2: customerID hatao
if 'customerID' in df.columns:
    df.drop('customerID', axis=1, inplace=True)

# Fix 3: get_dummies wala tareeqa
X = df.drop('Churn', axis=1)
X = pd.get_dummies(X, drop_first=True).astype(float)
y = (df['Churn'] == 'Yes').astype(int)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

model = Pipeline([
    ('scale', StandardScaler()),
    ('lr', LogisticRegression(max_iter=1000))
])
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

print(f"\nAccuracy: {accuracy_score(y_test, y_pred)*100:.2f}%")
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nReport:\n", classification_report(y_test, y_pred))

/tmp/ipykernel_48/2153295396.py:11: DeprecationWarning: Use dataset_load() instead of load_dataset(). load_dataset() will be removed in a future version.
  df = kagglehub.load_dataset(


Data Load Ho Gya! Shape: (7043, 21)

Accuracy: 80.70%

Confusion Matrix:
 [[925 110]
 [162 212]]

Report:
               precision    recall  f1-score   support

           0       0.85      0.89      0.87      1035
           1       0.66      0.57      0.61       374

    accuracy                           0.81      1409
   macro avg       0.75      0.73      0.74      1409
weighted avg       0.80      0.81      0.80      1409



## Week 2: Exploratory Data Analysis and Baseline Models
- Dataset: Telco Customer Churn, 7043 rows
- Cleaning: TotalCharges ko numeric kiya, 11 missing values ko 0 se fill kiya
- Encoding: pd.get_dummies with Pipeline and StandardScaler
- Model: Logistic Regression with 80/20 split
- Accuracy: ~80% (aap jo aayi hai wahi likh dena)
- Lesson: Accuracy se zyada Recall/AUC dekhna chahiye

## Week 2: Exploratory Data Analysis and Baseline Models
- Dataset: Telco Customer Churn, 7043 rows and 21 columns
- Cleaning: TotalCharges converted to numeric, 11 missing values filled with 0
- Encoding: One-hot encoding with pd.get_dummies and StandardScaler in Pipeline
- Split: 80/20 train-test with stratify
- Model: Logistic Regression
- Accuracy: 80.70%
- Key Metric: Confusion Matrix and Classification Report used for evaluation
- Lesson: Handling missing values properly improves accuracy and recall is important for churn